File: code/系統程式/textdata/diff0/diff.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char op;
    char *line;
} DiffOp;

char** split_lines(const char *content, int *line_count) {
    int count = 0;
    int len = strlen(content);
    for (int i = 0; i < len; i++) {
        if (content[i] == '\n') count++;
    }
    if (len > 0 && content[len-1] != '\n') count++;
    
    char **lines = malloc(count * sizeof(char*));
    char *copy = strdup(content);
    
    int idx = 0;
    char *start = copy;
    for (int i = 0; i < len; i++) {
        if (content[i] == '\n') {
            copy[i] = '\0';
            lines[idx++] = start;
            start = copy + i + 1;
        }
    }
    if (idx < count) lines[idx++] = start;
    *line_count = count;
    return lines;
}

void free_lines(char **lines, int count, char *buffer) {
    free(buffer);
    free(lines);
}

void diff_dp(char **A, int N, char **B, int M) {
    int **dp = malloc((N + 1) * sizeof(int*));
    for (int i = 0; i <= N; i++) {
        dp[i] = calloc(M + 1, sizeof(int));
    }

    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            if (strcmp(A[i-1], B[j-1]) == 0) {
                dp[i][j] = dp[i-1][j-1] + 1;
            } else {
                dp[i][j] = dp[i-1][j] > dp[i][j-1] ? dp[i-1][j] : dp[i][j-1];
            }
        }
    }

    int i = N, j = M;
    DiffOp *ops = malloc((N + M) * sizeof(DiffOp));
    int op_count = 0;

    while (i > 0 || j > 0) {
        if (i > 0 && j > 0 && strcmp(A[i-1], B[j-1]) == 0) {
            ops[op_count].op = ' ';
            ops[op_count].line = A[i-1];
            op_count++;
            i--; j--;
        } else if (i > 0 && (j == 0 || dp[i][j-1] < dp[i-1][j])) {
            ops[op_count].op = '-';
            ops[op_count].line = A[i-1];
            op_count++;
            i--;
        } else {
            ops[op_count].op = '+';
            ops[op_count].line = B[j-1];
            op_count++;
            j--;
        }
    }

    for (int k = op_count - 1; k >= 0; k--) {
        printf("%c %s\n", ops[k].op, ops[k].line);
    }

    free(ops);
    for (int i = 0; i <= N; i++) free(dp[i]);
    free(dp);
}

char* read_file(const char *filename) {
    FILE *f = fopen(filename, "r");
    if (!f) {
        fprintf(stderr, "Cannot open file: %s\n", filename);
        return NULL;
    }
    
    fseek(f, 0, SEEK_END);
    int len = ftell(f);
    fseek(f, 0, SEEK_SET);
    
    char *content = malloc(len + 1);
    fread(content, 1, len, f);
    content[len] = '\0';
    fclose(f);
    return content;
}

int main(int argc, char **argv) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <fileA> <fileB>\n", argv[0]);
        return 1;
    }

    char *contentA = read_file(argv[1]);
    char *contentB = read_file(argv[2]);
    
    if (!contentA || !contentB) {
        free(contentA);
        free(contentB);
        return 1;
    }

    int lineA, lineB;
    char **linesA = split_lines(contentA, &lineA);
    char **linesB = split_lines(contentB, &lineB);

    diff_dp(linesA, lineA, linesB, lineB);

    free_lines(linesA, lineA, contentA);
    free_lines(linesB, lineB, contentB);
    return 0;
}

File: code/系統程式/textdata/diff0/diff.rs
use std::env;
use std::fs;
use std::process;

// 定義與 C 語言相同的結構體，使用生命週期 'a 來借用字串切片
struct DiffOp<'a> {
    op: char,
    line: &'a str,
}

// 動態規劃實作 Diff 演算法
fn diff_dp<'a>(a: &[&'a str], b: &[&'a str]) {
    let n = a.len();
    let m = b.len();

    // 建立二維的 DP 表格並初始化為 0
    // 相當於 C 的: int **dp = calloc(...)
    let mut dp = vec![vec![0; m + 1]; n + 1];

    // 填寫 DP 表格 (Longest Common Subsequence)
    for i in 1..=n {
        for j in 1..=m {
            if a[i - 1] == b[j - 1] {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = dp[i - 1][j].max(dp[i][j - 1]);
            }
        }
    }

    let mut i = n;
    let mut j = m;
    let mut ops = Vec::new();

    // 回溯尋找操作路徑
    while i > 0 || j > 0 {
        if i > 0 && j > 0 && a[i - 1] == b[j - 1] {
            ops.push(DiffOp { op: ' ', line: a[i - 1] });
            i -= 1;
            j -= 1;
        } else if i > 0 && (j == 0 || dp[i][j - 1] < dp[i - 1][j]) {
            ops.push(DiffOp { op: '-', line: a[i - 1] });
            i -= 1;
        } else {
            ops.push(DiffOp { op: '+', line: b[j - 1] });
            j -= 1;
        }
    }

    // 因為回溯是從後往前，所以我們需要反轉輸出
    for op in ops.iter().rev() {
        println!("{} {}", op.op, op.line);
    }
}

// 讀取檔案的輔助函式，整合了錯誤處理
fn read_file(filename: &str) -> String {
    match fs::read_to_string(filename) {
        Ok(content) => content,
        Err(err) => {
            eprintln!("Cannot open file: {} ({})", filename, err);
            process::exit(1);
        }
    }
}

fn main() {
    // 解析命令列參數
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: {} <fileA> <fileB>", args[0]);
        process::exit(1);
    }

    // 讀取檔案內容
    // 相當於 C 中的 contentA 和 contentB
    let content_a = read_file(&args[1]);
    let content_b = read_file(&args[2]);

    // Rust 的 .lines() 會自動以換行符號分割字串，且不會產生額外的記憶體拷貝
    // 相當於 C 中的 split_lines 函式
    let lines_a: Vec<&str> = content_a.lines().collect();
    let lines_b: Vec<&str> = content_b.lines().collect();

    // 執行演算法
    diff_dp(&lines_a, &lines_b);

    // Rust 變數離開作用域時會自動釋放記憶體，不需要像 C 一樣呼叫 free_lines()
}